### PR TITLE
ASA validation; Deployer logic moved into separate file

### PR DESCRIPTION
**Pull request by Invertisment**

### CODE REVIEW

File: packages/algob/src/lib/validation-errors.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import * as z from 'zod';
+import { ZodErrorCode, ZodSuberror } from "zod/lib/src/ZodError";
+
+function parseZodSubError (e: ZodSuberror, indent: number): string {
```
Comment (robert-zaremba) [inline packages/algob/src/lib/validation-errors.ts:4]: Let's put this file next to the `errors-list.ts` file (in same for)

Comment (Invertisment) [reply packages/algob/src/lib/validation-errors.ts]: Moved.

File: packages/algob/src/types-input.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import * as z from 'zod';
+
+export const AddressSchema = z.string();
```
Comment (robert-zaremba) [inline packages/algob/src/types-input.ts:3]: Do we need this?

File: packages/algob/src/types-input.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    path: ['unitName']
+  })
+  .refine(o => (!o.url || (o.url && (o.url.length <= 32))), {
+    message: "URL must not be longer than 32 bytes",
+    path: ['url']
+  })
+  .refine(o => (!o.metadataHash || (o.metadataHash && (o.metadataHash.length <= 32))), {
+    message: "Metadata Hash must not be longer than 32 bytes",
+    path: ['metadataHash']
+  });
```
Comment (robert-zaremba) [inline packages/algob/src/types-input.ts:34]: Maybe we can have one define with many `if` inside. Same amount big code and maybe even better readability.

Comment (Invertisment) [reply packages/algob/src/types-input.ts:34]: I can have one large blob but they will return the same objects anyway.

File: packages/algob/src/types-input.ts (hunk 2 above)
Comment (Invertisment) [reply packages/algob/src/types-input.ts:3]: This is how address is decoded. 
It's used four times in the ASA definition:
```
  manager: AddressSchema.optional(),
  reserve: AddressSchema.optional(),
  freeze: AddressSchema.optional(),
  clawback: AddressSchema.optional()
```

File: packages/algob/src/types-input.ts (hunk 3 above)
Comment (Invertisment) [reply packages/algob/src/types-input.ts:34]: It's not possible with this method.
It accepts two parameters and it doesn't do whole object check.

https://github.com/vriad/zod/#custom-validation

File: packages/algob/src/lib/asa.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import { BuilderError } from "../internal/core/errors";
+import { ERRORS } from "../internal/core/errors-list";
+import { parseZodError } from "../internal/core/validation-errors";
+import { ASADescription } from "../types";
+import { ASADescriptionSchema } from "../types-input";
+
+export function parseASADef (obj: Object, filename: string): ASADescription {
+  try {
+    const parsed = ASADescriptionSchema.parse(obj);
+    if (parsed.defaultFrozen === undefined) {
```
[the remark this answers is not in the document]
Comment (Invertisment) [reply packages/algob/src/lib/asa.ts:13]: Renamed

File: packages/algob/src/types-input.ts (hunk 2 above)
Comment (robert-zaremba) [reply packages/algob/src/types-input.ts:3]: OK, thanks.

**Pull request merged**